src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java

```diff
@@ -0,0 +1,216 @@
+    return this.tableName;
+  }
+
+  /** @return List of column names specified in COPY statement, if provided. */
+  public List<String> getCopyColumnNames() {
+    return this.copyColumnNames;
+  }
+
+  /** @return Format type specified in COPY statement, if provided. */
+  public String getFormatType() {
+    return this.formatType;
+  }
+
+  /** @return Null string specified in COPY statement, if provided. */
+  public String getNullString() {
+    return this.nullString;
+  }
+
+  /** @return Delimiter character specified in COPY statement, if provided. */
+  public char getDelimiterChar() {
+    return this.delimiterChar;
+  }
+
+  /** @return Escape character specified in COPY statement, if provided. */
+  public char getEscapeChar() {
+    return this.escapeChar;
+  }
+
+  /** @return Quote character specified in COPY statement, if provided. */
+  public char getQuoteChar() {
+    return this.quoteChar;
+  }
+
+  /** @return True if copy data contains a header, false otherwise. */
+  public boolean hasHeader() {
+    return this.hasHeader;
+  }
+
+  public int getFormatCode() {
+    return this.formatCode;
+  }
+
+  public MutationBuilder getMutationBuilder() {
+    return this.mutationBuilder;
+  }
+
+  private void verifyCopyColumns() throws SQLException {
+    if (this.copyColumnNames.size() > this.tableColumns.size()) {
+      throw new SQLException("Number of copy columns provided exceed table column count");
+    }
+    LinkedHashMap<String, TypeCode> tempTableColumns = new LinkedHashMap<>();
+    // Verify that every copy column given is a valid table column name
+    for (String copyColumn : this.copyColumnNames) {
+      if (!this.tableColumns.containsKey(copyColumn)) {
+        throw new SQLException(
+            "Column \"" + copyColumn + "\" of relation \"" + this.tableName + "\" does not exist");
+      }
+      // Update table column ordering with the copy column ordering
+      tempTableColumns.put(copyColumn, tableColumns.get(copyColumn));
+    }
+    this.tableColumns = tempTableColumns;
+  }
+
+  private static TypeCode parseSpannerDataType(String columnType) {
+    // Eliminate size modifiers in column type (e.g. STRING(100), BYTES(50), etc.)
+    int index = columnType.indexOf("(");
+    columnType = (index > 0) ? columnType.substring(0, index) : columnType;
+    TypeCode type = TypeCode.valueOf(columnType);
+    if (type == null) {
+      throw new IllegalArgumentException(
+          "Unrecognized or unsupported column data type: " + columnType);
+    }
+    return type;
+  }
+
+  private void queryInformationSchema() throws SQLException {
+    Map<String, TypeCode> tableColumns = new LinkedHashMap<>();
+    PreparedStatement statement =
+        this.connection.prepareStatement(
+            "SELECT "
+                + COLUMN_NAME
+                + ", "
+                + SPANNER_TYPE
+                + " FROM information_schema.columns WHERE table_name = ?");
+    statement.setString(1, this.tableName);
```

Comment:
Note that this requires that the table name is correctly handled by the parser and folded to lower case unless it was quoted:
1. Spangres will store the table name in lower-case, unless it was quoted when it was created.
2. If the user specifies the table name in the copy statement in upper case or mixed case without quoting, it will not be found by this query and cause an exception, unless the parser correctly applies PostgreSQL case folding of identifiers.

Reply:
I think we will enforce this in the Parser.  If not, I'll update this here.